### PR TITLE
feat(catalog-backend): move filters into their own query key

### LINKED ISSUE #2910: Move Catalog REST API filters into their own query parameter
## Feature Suggestion

All query parameters are treated as filters when querying the catalog right now, implemented [here](https://github.com/spotify/backstage/blob/7ff2f1c0f0c63b8be5712a56569b83743b79f980/plugins/catalog-backend/src/service/router.ts#L128). We should move filters into a dedicated query parameter such as `?filter=` or `?q=` in order to make space for things like pagination and sorting parameters.

## Possible Implementation

Either move in the filters mostly as is, something like `?filter=metadata.name=my-component&filter=kind=component`. Or encode the same data in something like JSON so that we can evolve it a bit easier in the future: `?filter={"metadata":{"name":"my-component"},"kind":"component"}`

## Context

Settling a reasonably stable version of the catalog API, which we're able to evolve over time.

Also a previous discussion on Discord: https://discordapp.com/channels/687207715902193673/687207715902193679/759482223388393473

### PR DESCRIPTION
**Pull request by freben**
Fixes #2910 

You may want to ignore whitespace while reviewing.

### CODE REVIEW

File: plugins/auth-backend/src/lib/catalog/CatalogIdentityClient.ts [hunk 1]
```diff
@@ -42,15 +42,15 @@ export class CatalogIdentityClient {
    * Throws a NotFoundError or ConflictError if 0 or multiple users are found.
    */
   async findUser(query: UserQuery): Promise<UserEntity> {
-    const params = new URLSearchParams();
-    params.append('kind', 'User');
-
+    const conditions = ['kind=user'];
     for (const [key, value] of Object.entries(query.annotations)) {
-      params.append(`metadata.annotations.${key}`, value);
+      conditions.push(`metadata.annotations.${key}=${value}`);
```
Comment (Rugvip) [inline plugins/auth-backend/src/lib/catalog/CatalogIdentityClient.ts:47]: `URLSearchParams` handles encoding, need to url encode both the key and the value now.

File: plugins/catalog-backend/src/database/types.ts [hunk 2]
```diff
@@ -153,7 +153,10 @@ export type Database = {
     matchingGeneration?: number,
   ): Promise<DbEntityResponse>;
 
-  entities(tx: unknown, filters?: EntityFilters): Promise<DbEntityResponse[]>;
+  entities(
+    tx: unknown,
+    filters?: EntityFilters | EntityFilters[],
```
Comment (Rugvip) [inline plugins/catalog-backend/src/database/types.ts:158]: This is more or less an internal interface that doesn't need to be super easy to type. I'd go with only `EntityFilters[]` instead.

File: plugins/catalog-backend/src/service/filterQuery.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import { EntityFilters } from '../database';
+
+export function translateQueryToEntityFilters(
+  query: Record<string, any>,
+): EntityFilters[] {
+  if (!query.filter) {
+    return [];
+  }
+
+  const filterStrings = Array.isArray(query.filter)
```
Comment (Rugvip) [inline plugins/catalog-backend/src/service/filterQuery.ts:27]: `[query.filter].flat()` would be an option here

**Review by Rugvip — COMMENTED** [review]
🎉

File: plugins/catalog-backend/src/service/filterQuery.ts (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/src/service/filterQuery.ts:27]: Aw yiss, I forget

File: plugins/catalog-backend/src/database/types.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/database/types.ts:158]: Sure.

I also thought there's a little too much logic spread around, and EntityFilters could be an interface that exposes `matchNulls`, `matchIn` etc. An implementing class of that could have static builder methods that deal with the parsing of the query string format, or translating from the object format. That would concentrate all of this better and remove a fair bit of code from the db layer.

File: plugins/auth-backend/src/lib/catalog/CatalogIdentityClient.ts (hunk 1 above)
Comment (freben) [reply plugins/auth-backend/src/lib/catalog/CatalogIdentityClient.ts:47]: God call, had it on my mind but forgot

**Comment by freben** [comment]
Updated according to comments.

**Review by Rugvip — APPROVED** [review]

**Pull request merged**